### PR TITLE
Support for immediate saving to reduce ram usage

### PR DESCRIPTION
**Pull request by Kaihui-intel**
### Accuracy
scheme /(opt-125m,) | format | RTN | iter>0
-- | -- | -- | --
W4A16 | auto_round | 0.2882 | 0.3526
W2A16 | auto_round |   | 0.1657
W3A16 | auto_round |   | 0.3247
W8A16 | auto_round |   | 0.3784
bit s group_size 32 | auto_round | 0.3749 | 0.3679
bit s group_size 32 | auto_gptq | 0.3747 | 0.3658
bit s group_size 32 | auto_awq | 0.3749 | 0.3646

https://github.com/intel/auto-round/issues/788

### Memory
memory check 
Qwen2.5-7B-Instruct-w4g32 RTN auto_round
mprof peak
`16659.441MiB`->`9200.250MiB` ~55%

### Time
quantization and saving time
opt-125m
branch |  RTN | iter>0
-- | -- | -- |
cur branch | 67s | 78s
main branch | 50s | 88s

Qwen2.5-7B-Instruct
branch |  RTN | iter>0
-- | -- | -- |
cur branch | 4min4s | 18min10s
main branch | 3min54s | 17min53s

 immediate pacing and saving now only support formats[0]

### CODE REVIEW

**Comment by wenhuach21** [comment]
Thanks for the great work! Could you check the maximum RAM usage to see whether it has been reduced significantly, as expected?

File: auto_round/compressors/base.py [hunk 1]
```diff
@@ -2797,6 +2810,223 @@ def _quantize_blocks(
 
         clear_memory()
 
+    def _save_block_immediate(self, m, last_group=False):
+        import json
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: better move to other file as this file is too long

File: auto_round/compressors/base.py [hunk 2]
```diff
@@ -1271,6 +1275,12 @@ def _quantize_layer_via_rtn(self, name: str) -> None:
         else:
             set_module(self.model, name, m)
 
+        if self.save_block_immediate:
+            all_to_quantized_module_names = [n for n, m in self.model.named_modules() if check_to_quantized(m)]
+            last_module = (len(all_to_quantized_module_names) == 0) or (name == all_to_quantized_module_names[-1])
+            m = get_module(self.model, name)
```
Comment (wenhuach21) [inline auto_round/compressors/base.py:1277]: please test lm_head quantization and vlm quantization.
better add uts.

File: auto_round/utils/common.py [hunk 3]
```diff
@@ -297,3 +297,17 @@ def get_reciprocal(tensor):
     else:
         tensor = torch.where(torch.abs(tensor) < 1e-30, 0, tensor)
     return torch.where(tensor != 0, 1 / tensor, torch.zeros_like(tensor))
+
+
+def is_meta_model(model: torch.nn.Module) -> bool:
+    is_meta = False
```
Comment (wenhuach21) [inline auto_round/utils/common.py]: I remember we have some functions to do the same thing, @n1ck-guo

File: auto_round/compressors/utils.py [hunk 4]
```diff
@@ -1027,3 +1027,239 @@ def reset_params(inputs):
     """
     if "use_cache" in inputs.keys():  # Not storing kv cache
         inputs["use_cache"] = False
+
+
+def save_block_immediate(rounder, m, name=None, last_group=False):
+    """
```
Comment (wenhuach21) [inline auto_round/compressors/utils.py:1037]: type annotation

File: auto_round/compressors/base.py (hunk 2 above)
Comment (Kaihui-intel) [reply auto_round/compressors/base.py:1277]: Qwen3-8B-w4g32(lm_head) Qwen2-VL-2B-Instruct-w4g32(vlm) pass

Comment (wenhuach21) [reply auto_round/compressors/base.py:1277]: add uts in cpu and cuda

File: auto_round/compressors/base.py [hunk 5]
```diff
@@ -333,7 +334,10 @@ def __init__(
         self.lr_scheduler = lr_scheduler
         self.optimizer = self._get_optimizer(None)
         self.disable_opt_rtn = disable_opt_rtn
-        self.is_packing_immediate = False  # whether to pack the layer immediately after tuning
+
+        # Whether to pack the layer immediately after tuning
+        self.is_packing_immediate = kwargs.pop("is_packing_immediate", False)
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: Packing immediate is set automatically before. Have you handled this when exporting >1 formats? So it’s better not to set it in the API. Besides, as discussed, set save_immediate to True.
Another thing to verify is the time cost of save_immediate, have you measured the total quantization time comparing to main branch ?

File: auto_round/compressors/base.py [hunk 6]
```diff
@@ -333,7 +334,10 @@ def __init__(
         self.lr_scheduler = lr_scheduler
         self.optimizer = self._get_optimizer(None)
         self.disable_opt_rtn = disable_opt_rtn
-        self.is_packing_immediate = False  # whether to pack the layer immediately after tuning
+
+        # Whether to pack the layer immediately after tuning
+        self.immediate_packing = kwargs.pop("immediate_packing", False)
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: no need to expose this arg, automatically setting this is a better way

File: auto_round/compressors/base.py [hunk 7]
```diff
@@ -1726,6 +1739,38 @@ def _quantize_layers(self, layer_names: list, layer_inputs: dict) -> None:
             q_layer_input = q_layer_inputs.get(layer_name, None) if q_layer_inputs is not None else None
             q_layer_input = to_device(q_layer_input, self.cache_device)
             quant_layer(layer_name, layer_input, q_layer_input, device=self.device)
+            if self.immediate_packing:
+                from auto_round.export import PACKING_LAYER_WITH_FORMAT
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: wrapper it  as a function

File: auto_round/compressors/base.py (hunk 6 above)
Comment (wenhuach21) [inline auto_round/compressors/base.py]: where is the code to set it to False for fake format or multiple formats

Comment (Kaihui-intel) [reply auto_round/compressors/base.py]: https://github.com/intel/auto-round/blob/5375be6db65ead23845551d4ee449b59112d2272/auto_round/compressors/base.py#L1546-L1562

File: auto_round/compressors/utils.py [hunk 8]
```diff
@@ -1031,3 +1031,241 @@ def reset_params(inputs):
...
+            for free_module_key in free_module_state_dict:
+                free_module_key_full_name = f"{free_module_name}.{free_module_key}"
+                if free_module_key_full_name not in already_saved_name:
+                    free_flag = False
+            if free_flag:
+                free_module.to("meta")
+                del rounder._current_shard_tensors[param]
+        rounder._current_shard_tensors = OrderedDict()
+        rounder._current_shard_size = 0
+        clear_memory()
```
Comment (wenhuach21) [inline auto_round/compressors/utils.py:1138]: use gc.collect is enough? empty_cache causes large overhead

Comment (Kaihui-intel) [reply auto_round/compressors/utils.py:1138]: I have checked that it can effectively free up tensor memory.

File: test/test_cpu/test_autoround.py [hunk 9]
```diff
@@ -808,6 +808,27 @@ def test_attention_mask_via_tokenize_in_dataset(self):
...
+        quantized_model_path = self.save_folder
+        autoround = AutoRound(
+            model,
+            tokenizer,
+            bits=bits,
+            group_size=group_size,
+            iters=2,
+            seqlen=10,
+            dataset=self.llm_dataloader,
+            is_packing_immediate=True,
```
Comment (wenhuach21) [inline test/test_cpu/test_autoround.py]: why this value could still be passed to AutoRound, if there is no warning , then you need to have  a check

File: auto_round/compressors/base.py [hunk 10]
```diff
@@ -336,7 +337,11 @@ def __init__(
         self.lr_scheduler = lr_scheduler
         self.optimizer = self._get_optimizer(None)
         self.disable_opt_rtn = disable_opt_rtn
-        self.is_packing_immediate = False  # whether to pack the layer immediately after tuning
+
+        # Whether to pack the layer immediately after tuning
+        self.immediate_packing = False
+        self.immediate_saving = False
+        self.low_cpu_mem_usage = kwargs.pop("low_cpu_mem_usage", False)
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: move the code in front of the following, otherwise, there is a warning
~~~python
        if kwargs:
            logger.warning(f"unrecognized keys {list(kwargs.keys())} were passed. Please check them.")
~~~

Comment (wenhuach21) [inline auto_round/compressors/base.py]: add comments for "low_cpu_mem_usage" in AutoRound API and base compressor  like other kwargs

File: test/test_cpu/test_autoround.py (hunk 9 above)
Comment (Kaihui-intel) [reply test/test_cpu/test_autoround.py]: removed. use low_cpu_mem_usage

File: auto_round/compressors/base.py [hunk 11]
```diff
@@ -144,6 +144,7 @@ def __init__(
         batch_size: int = 8,
         gradient_accumulate_steps: int = 1,
         low_gpu_mem_usage: bool = False,
+        low_cpu_mem_usage: bool = False,
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: move to the last

File: auto_round/autoround.py [hunk 12]
```diff
@@ -78,6 +78,7 @@ def __new__(
         batch_size: int = 8,
         gradient_accumulate_steps: int = 1,
         low_gpu_mem_usage: bool = False,
+        low_cpu_mem_usage: bool = False,
```
Comment (wenhuach21) [inline auto_round/autoround.py]: move to the last in case users pass with tuple

**Review by wenhuach21 — APPROVED** [review]
